libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/exclusive_scan.hpp

```diff
@@ -0,0 +1,662 @@
+    /// invoked with an execution policy object of type \a sequenced_policy
+    /// execute in sequential order in the calling thread.
+    ///
+    /// The reduce operations in the parallel \a exclusive_scan algorithm
+    /// invoked with an execution policy object of type \a parallel_policy
+    /// or \a parallel_task_policy are permitted to execute in an unordered
+    /// fashion in unspecified threads, and indeterminately sequenced
+    /// within each thread.
+    ///
+    /// \returns  The \a exclusive_scan algorithm returns a
+    ///           \a hpx::future<O> if
+    ///           the execution policy is of type
+    ///           \a sequenced_task_policy or
+    ///           \a parallel_task_policy and
+    ///           returns \a O otherwise.
+    ///           The \a exclusive_scan algorithm returns the output iterator
+    ///           to the element in the destination range, one past the last
+    ///           element copied.
+    ///
+    /// \note   GENERALIZED_NONCOMMUTATIVE_SUM(+, a1, ..., aN) is defined as:
+    ///         * a1 when N is 1
+    ///         * GENERALIZED_NONCOMMUTATIVE_SUM(+, a1, ..., aK)
+    ///           + GENERALIZED_NONCOMMUTATIVE_SUM(+, aM, ..., aN)
+    ///           where 1 < K+1 = M <= N.
+    ///
+    /// The difference between \a exclusive_scan and \a inclusive_scan is that
+    /// \a inclusive_scan includes the ith input element in the ith sum.
+    ///
+    template <typename ExPolicy, typename Rng, typename O, typename T,
+        typename Op>
+    typename util::detail::algorithm_result<ExPolicy, O>::type exclusive_scan(
+        ExPolicy&& policy, Rng&& rng, O dest, T init, Op&& op);
+}}    // namespace hpx::ranges
+#else
+
+#include <hpx/config.hpp>
+#include <hpx/algorithms/traits/projected_range.hpp>
+#include <hpx/execution/algorithms/detail/predicates.hpp>
+#include <hpx/executors/execution_policy.hpp>
+#include <hpx/functional/tag_fallback_dispatch.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <hpx/parallel/algorithms/exclusive_scan.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace ranges {
+    HPX_INLINE_CONSTEXPR_VARIABLE struct exclusive_scan_t final
+      : hpx::functional::tag_fallback<exclusive_scan_t>
+    {
+    private:
+        // clang-format off
+        template <typename InIter, typename Sent, typename OutIter,
+             typename T, typename Op = std::plus<T>,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<InIter>::value &&
+                hpx::traits::is_sentinel_for<Sent, InIter>::value &&
+                hpx::traits::is_iterator<OutIter>::value
+            )>
+        // clang-format on
+        friend OutIter tag_fallback_dispatch(hpx::ranges::exclusive_scan_t,
```

Comment:
Should we return an `in_out_result<InIter, OutIter>` here? It might be useful to return the iterator that corresponds to the position the sentinel refers to.

Reply:
Oh ok sure, I'll make the change in inclusive_scan too.